openmls/src/ciphersuite/mod.rs

```diff
@@ -538,6 +549,203 @@ impl Signature {
+
+            // The verification algorithm expects the scalars to be 32 bytes
+            // long, buffered with zeroes.
+            let mut padded_scalar = vec![0u8; P256_SCALAR_LENGTH - scalar_length];
+            padded_scalar.append(&mut scalar);
+
+            Ok(padded_scalar)
+        }
+
+        let mut signature_bytes = self.value.as_slice();
+
+        // Check header bytes:
+        // 1 byte SEQUENCE tag should be 0x30
+        let sequence_tag = signature_bytes
+            .read_u8()
+            .map_err(|_| SignatureError::DecodingError)?;
+        if sequence_tag != SEQUENCE_TAG {
+            return Err(SignatureError::DecodingError);
+        };
+
+        // At most 1 byte encoding the length of the scalars (short form DER
+        // length encoding). Length has to be encoded in the short form, as we
+        // expect the length not to exceed the maximum length of 70: Two times
+        // at most 32 (scalar value) + 1 byte integer tag + 1 byte length tag +
+        // at most 1 byte to indicating that the integer is unsigned.
+        let length = signature_bytes
+            .read_u8()
+            .map_err(|_| SignatureError::DecodingError)? as usize;
+        if length > 2 * (P256_SCALAR_LENGTH + 3) {
+            return Err(SignatureError::DecodingError);
+        }
+
+        // The remaining bytes should be equal to the encoded length.
+        if signature_bytes.len() != length {
+            return Err(SignatureError::DecodingError);
+        }
+
+        let mut r = decode_scalar(&mut signature_bytes)?;
+        let mut s = decode_scalar(&mut signature_bytes)?;
+
+        // After reading the whole signature, nothing should be left.
+        if !signature_bytes.is_empty() {
+            return Err(SignatureError::DecodingError);
+        }
+
+        let mut out = Vec::with_capacity(2 * P256_SCALAR_LENGTH);
+        out.append(&mut r);
+        out.append(&mut s);
+        Ok(out)
+    }
+
+    /// This function DER encodes a given ECDSA signature consisting of bytes
+    /// representing the concatenated scalars. If the encoding fails, it will
+    /// throw a `SignatureError`.
+    pub(crate) fn der_encode(raw_signature: &[u8]) -> Result<Self, SignatureError> {
+        // A small function to DER encode single scalar.
+        fn encode_scalar(mut scalar: &[u8]) -> Result<Vec<u8>, SignatureError> {
+            // We first prepare the scalar DER encoded as a uint.
+            let mut scalar_uint: Vec<u8> = Vec::with_capacity(P256_SCALAR_LENGTH + 1);
+
+            // Remove prepending zeros
+            let mut msb = scalar
+                .read_u8()
+                .map_err(|_| SignatureError::EncodingError)?;
+            while msb == 0x00 {
+                msb = scalar
+                    .read_u8()
+                    .map_err(|_| SignatureError::EncodingError)?;
+            }
+
+            // If the most significant bit is 1, we have to prepend 0x00 to indicate
```

Comment:
So we first remove all `0x00` and then add one again? You could peek the next byte. But this might end up being more complex so i guess this is fine.